src/decapodes_it.jl

```diff
@@ -0,0 +1,272 @@
+        end
+        return res_var
+      end
+      # TODO: Just for now assuming we have 2 or more terms
+      decapodes.Mult(ts) => begin
+        multiplicands  = [!(t,d,syms) for t in ts]
+        res_var = add_part!(d, :Var, type=:infer, name=:mult)
+        m1,m2 = multiplicands[1:2]
+        add_part!(d, :Op2, proj1=m1, proj2=m2, res=res_var, op2=Symbol("*"))
+        for m in multiplicands[3:end]
+          m1 = res_var
+          m2 = m
+          res_var = add_part!(d, :Var, type=:infer, name=:mult)
+          add_part!(d, :Op2, proj1=m1, proj2=m2, res=res_var, op2=Symbol("*"))
+        end
+        return res_var
+      end
+      decapodes.Tan(t) => begin 
+        # TODO: this is creating a spurious variable with the same name
+        txv = add_part!(d, :Var, type=:infer)
+        tx = add_part!(d, :TVar, incl=txv)
+        tanop = add_part!(d, :Op1, src=!(t,d,syms), tgt=txv, op1=DerivOp)
+        return txv #syms[x[1]]
+      end
+      _ => throw("Inline type judgements not yet supported!")
+    end
+  end
+
+function eval_eq!(eq::decapodes.Eq, d::decapodes.AbstractDecapode, syms::Dict{Symbol, Int}, deletions::Vector{Int}) 
+  @match eq begin
+    decapodes.Eq(t1, t2) => begin
+      lhs_ref = reduce_term!(t1,d,syms)
+      rhs_ref = reduce_term!(t2,d,syms)
+
+      # Always let the a named variable take precedence 
+      # TODO: If we have variable to variable equality, we want
+      # some kind of way to check track of this equality
+      ref_pair = (t1, t2)
+      @match ref_pair begin
+        (decapodes.Var(a), decapodes.Var(b)) => return d
+        (t1, decapodes.Var(b)) => begin
+          lhs_ref, rhs_ref = rhs_ref, lhs_ref
+        end
+        _ => nothing
+      end
+
+      # Make rhs_ref equal to lhs_ref and adjust all its incidents
+
+      # Case rhs_ref is a Tan
+      # WARNING: Don't push to deletion here because all TanVars should have a 
+      # corresponding Op1. Pushing here would create a duplicate which breaks rem_parts!
+      for rhs in incident(d, rhs_ref, :incl)
+        d[rhs, :incl] = lhs_ref
+      end
+      # Case rhs_ref is a Op1
+      for rhs in incident(d, rhs_ref, :tgt)
+        d[rhs, :tgt] = lhs_ref
+        push!(deletions, rhs_ref)
+      end
+      # Case rhs_ref is a Op2
+      for rhs in incident(d, rhs_ref, :res)
+        d[rhs, :res] = lhs_ref
+        push!(deletions, rhs_ref)
+      end
+      # Case rhs_ref is a Plus
+      # FIXME: this typeguard is a subsitute for refactoring into multiple dispatch
+      if isa(d, decapodes.SummationDecapode)
+        for rhs in incident(d, rhs_ref, :sum)
+          d[rhs, :sum] = lhs_ref
+          push!(deletions, rhs_ref)
+        end
+      end
+      # TODO: delete unused vars. The only thing stopping me from doing 
+      # this is I don't know if CSet deletion preserves incident relations
+      #rem_parts!(d, :Var, sort(deletions))
+    end
+  end
+  return d
+end
+
+function recognize_types(d::decapodes.AbstractNamedDecapode)
```

Comment:
Should SynacticModels be doing decapodes operations like `recognize_types`? What do you think the scope boundary is?